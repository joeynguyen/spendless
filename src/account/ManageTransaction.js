import React, { PropTypes } from 'react';
import { Modal, Button, Input } from 'react-bootstrap';

const ManageTransaction = ({ manageType = 'add', fields, toggleManageTransaction, pristine = true, doSubmit }) => {
  const { date, description, category, amount, notes } = fields;
  let componentTitle = 'Add Transaction';
  if (manageType === 'edit') {
    componentTitle = 'Edit Transaction';
  }

  return (
    <form onSubmit={doSubmit}>
      <Modal.Header closeButton>
        <Modal.Title>{ componentTitle }</Modal.Title>
      </Modal.Header>

      <Modal.Body>
        <div className="row">
          <div className="col-xs-12">
            <Input
              type="text"
              label="Description"
              bsStyle={description.touched && description.invalid ? 'error' : null}
              help={description.touched ? description.error : ''}
              {...description}
            />
            <Input
              type="date"
              label="Date"
              bsStyle={date.touched && date.invalid ? 'error' : null}
              help={date.touched ? date.error : ''}
              {...date}
            />
            <Input
              type="text"
              label="Category"
              bsStyle={category.touched && category.invalid ? 'error' : null}
              help={category.touched ? category.error : ''}
              {...category}
            />
            <Input
              type="text"
              addonBefore="$"
              label="Amount"
              bsStyle={amount.touched && amount.invalid ? 'error' : null}
              help={amount.touched ? amount.error : ''}
              {...amount}
            />
            <Input
              type="textarea"
              label="Notes"
              placeholder="Add notes for this transaction"
              {...notes}
            />
          </div>
        </div>
      </Modal.Body>

      <Modal.Footer>
        <div className="form-group">
          <Button
            bsStyle="primary"
            type="submit"
            disabled={pristine}
          >Save</Button>
          {' '}
          <Button id="cancel-manage-transaction" onClick={toggleManageTransaction}>Cancel</Button>
        </div>
      </Modal.Footer>
    </form>
  );
};
ManageTransaction.propTypes = {
  manageType: PropTypes.string.isRequired,
  toggleManageTransaction: PropTypes.func.isRequired,
  fields: PropTypes.object.isRequired,
  pristine: PropTypes.bool.isRequired,
  doSubmit: PropTypes.func.isRequired,
};

export default ManageTransaction;
