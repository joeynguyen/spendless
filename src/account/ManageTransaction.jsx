import PropTypes from 'prop-types';
import React from 'react';
import { Button, DatePicker, Form, Input, Modal } from 'antd';

const FormItem = Form.Item;
const { TextArea } = Input;

const ManageTransaction = ({ form, manageType = 'add', manageTransactionVisible, toggleManageTransaction, doSubmit }) => {
  const { getFieldDecorator } = form;
  const componentTitle = (manageType === 'edit') ? 'Edit Transaction' : 'Add Transaction';
  const modalFooter = [(
    <Button
      key="confirm"
      type="primary"
      htmlType="submit"
      form="manage-transaction"
    >Save</Button>
  ), (
    <Button
      key="cancel"
      id="cancel-manage-transaction"
      size="large"
      onClick={toggleManageTransaction}
    >Cancel</Button>
  )];

  return (
    <Modal
      closable
      maskClosable={false}
      visible={manageTransactionVisible}
      onCancel={toggleManageTransaction}
      title={componentTitle}
      footer={modalFooter}
    >
      <Form id="manage-transaction" onSubmit={doSubmit} layout="vertical">
        <FormItem label="Description">
          {getFieldDecorator('description', {
            rules: [{ required: true, message: 'Enter a description' }],
          })(
            <Input />
          )}
        </FormItem>

        <FormItem label="Date">
          {getFieldDecorator('date', {
            rules: [{ required: true, message: 'Enter a date' }],
          })(
            <DatePicker />
          )}
        </FormItem>

        <FormItem label="Category">
          {getFieldDecorator('category', {
            rules: [{ required: true, message: 'Enter a category' }],
          })(
            <Input />
          )}
        </FormItem>

        <FormItem label="Amount">
          {getFieldDecorator('amount', {
            rules: [{ required: true, message: 'Enter an amount' }],
          })(
            <Input addonBefore="$" />
          )}
        </FormItem>

        <FormItem label="Notes">
          {getFieldDecorator('notes', {})(
            <TextArea rows={4} placeholder="Add notes for this transaction"/>
          )}
        </FormItem>
      </Form>
    </Modal>
  );
};

ManageTransaction.propTypes = {
  form: PropTypes.object.isRequired,
  initialValues: PropTypes.object.isRequired,
  manageTransactionVisible: PropTypes.bool.isRequired,
  manageType: PropTypes.string.isRequired,
  toggleManageTransaction: PropTypes.func.isRequired,
  doSubmit: PropTypes.func.isRequired,
};

export default ManageTransaction;

