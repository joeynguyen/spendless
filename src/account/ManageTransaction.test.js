import expect from 'expect';
import React from 'react';
import { mount } from 'enzyme';
import ManageTransaction from './ManageTransaction.jsx';

function setup() {
  // Need to create spies because if we use expect.spyOn on actual functions, we get this
  // error when test is run: "TypeError: Cannot assign to read only property 'doSubmit'"
  const mockOnSubmit = expect.createSpy();
  const mockOnCancel = expect.createSpy();
  const props = {
    manageType: 'edit',
    pristine: true,
    toggleManageTransaction: mockOnCancel,
    doSubmit: mockOnSubmit,
    fields: {
      amount: {
        name: 'amount',
        value: '12.34',
      },
      category: {
        name: 'category',
        value: 'Auto',
      },
      date: {
        name: 'date',
        value: '2014-07-25',
      },
      description: {
        name: 'description',
        value: 'This is a description',
      },
      notes: {
        name: 'notes',
        value: 'Add notes here',
      }
    },
  };

  return mount(<ManageTransaction {...props} />);
}
describe('ManageTransaction', () => {
  const component = setup();

  it('renders correct modal title', () => {
    expect(component.find('.modal-title').text()).toEqual('Edit Transaction');
  });

  describe('renders all the fields with correct values', () => {
    it('amount', () => {
      const amountField = component.find('input[name="amount"]');
      expect(amountField.length).toBe(1);
      expect(amountField.prop('value')).toBe('12.34');
    });

    it('category', () => {
      const categoryField = component.find('input[name="category"]');
      expect(categoryField.length).toBe(1);
      expect(categoryField.prop('value')).toBe('Auto');
    });

    it('date', () => {
      const dateField = component.find('input[name="date"]');
      expect(dateField.length).toBe(1);
      expect(dateField.prop('value')).toBe('2014-07-25');
    });

    it('description', () => {
      const descField = component.find('input[name="description"]');
      expect(descField.length).toBe(1);
      expect(descField.prop('value')).toBe('This is a description');
    });

    it('notes', () => {
      const notesField = component.find('textarea[name="notes"]');
      expect(notesField.length).toBe(1);
      expect(notesField.prop('value')).toBe('Add notes here');
    });
  });

  describe('has submit button that submits', () => {
    const submitBtn = component.find('button[type="submit"]');

    it('has submit button', () => {
      expect(submitBtn.length).toBe(1);
    });

    it('button initial state is disabled', () => {
      // test for existence b/c the prop either exists or doesn't exist,
      // it doesn't have a true or false value
      expect(submitBtn.prop('disabled')).toExist();
    });

    it('it submits on submit', (done) => {
      const spy = expect.spyOn(component.props(), 'doSubmit');

      // We can't test clicking on submit button to perform 'doSubmit' until
      // this bug is resolved - https://github.com/airbnb/enzyme/issues/308
      expect(spy.calls.length).toEqual(0);
      component.find('form').simulate('submit');
      expect(spy.calls.length).toEqual(1);

      spy.restore();
      done();
    });
  });

  describe('has cancel button that cancels', () => {
    const cancelBtn = component.find('button#cancel-manage-transaction');

    it('cancel button exists', () => {
      expect(cancelBtn.length).toBe(1);
    });

    it('it cancels on click', (done) => {
      const spy = expect.spyOn(component.props(), 'toggleManageTransaction');

      expect(spy.calls.length).toEqual(0);
      cancelBtn.simulate('click');
      expect(spy.calls.length).toEqual(1);

      spy.restore();
      done();
    });
  });
});
