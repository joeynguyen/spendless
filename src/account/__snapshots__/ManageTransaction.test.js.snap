// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageTransaction renders correctly 1`] = `
<form
  className="ant-form ant-form-vertical"
  id="manage-transaction"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item"
  >
    <div
      className="ant-form-item-label"
    >
      <label
        className="ant-form-item-required"
        htmlFor="description"
        onClick={[Function]}
        title="Description"
      >
        Description
      </label>
    </div>
    <div
      className="ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "description",
                "value": "This is a description",
              }
            }
            data-__meta={
              Object {
                "name": "description",
                "originalProps": Object {
                  "disabled": false,
                  "prefixCls": "ant-input",
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Enter a description",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Enter a description",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            disabled={false}
            id="description"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value="This is a description"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row"
  >
    <div
      className="ant-col-8"
    >
      <div
        className="ant-row ant-form-item"
      >
        <div
          className="ant-form-item-label"
        >
          <label
            className="ant-form-item-required"
            htmlFor="amount"
            onClick={[Function]}
            title="Amount"
          >
            Amount
          </label>
        </div>
        <div
          className="ant-form-item-control-wrapper"
        >
          <div
            className="ant-form-item-control has-success"
          >
            <span
              className="ant-form-item-children"
            >
              <span
                className="ant-input-group-wrapper"
              >
                <span
                  className="ant-input-wrapper ant-input-group"
                >
                  <span
                    className="ant-input-group-addon"
                  >
                    $
                  </span>
                  <input
                    className="ant-input"
                    data-__field={
                      Object {
                        "name": "amount",
                        "value": "12.34",
                      }
                    }
                    data-__meta={
                      Object {
                        "name": "amount",
                        "originalProps": Object {
                          "addonBefore": "$",
                          "disabled": false,
                          "prefixCls": "ant-input",
                          "type": "text",
                        },
                        "ref": null,
                        "rules": Array [
                          Object {
                            "message": "Enter an amount",
                            "required": true,
                          },
                        ],
                        "trigger": "onChange",
                        "validate": Array [
                          Object {
                            "rules": Array [
                              Object {
                                "message": "Enter an amount",
                                "required": true,
                              },
                            ],
                            "trigger": Array [
                              "onChange",
                            ],
                          },
                        ],
                        "valuePropName": "value",
                      }
                    }
                    disabled={false}
                    id="amount"
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    style={null}
                    type="text"
                    value="12.34"
                  />
                </span>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-col-12 ant-col-offset-4"
    >
      <div
        className="ant-row ant-form-item"
      >
        <div
          className="ant-form-item-label"
        >
          <label
            className="ant-form-item-required"
            htmlFor="date"
            onClick={[Function]}
            title="Date"
          >
            Date
          </label>
        </div>
        <div
          className="ant-form-item-control-wrapper"
        >
          <div
            className="ant-form-item-control has-success"
          >
            <span
              className="ant-form-item-children"
            >
              <span
                className="ant-calendar-picker"
                id="date"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              >
                <div
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                >
                  <input
                    className="ant-calendar-picker-input ant-input"
                    placeholder="Select date"
                    readOnly={true}
                    value="2014-07-25"
                  />
                  <i
                    className="anticon anticon-close-circle ant-calendar-picker-clear"
                    onClick={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className=""
                      data-icon="close-circle"
                      fill="currentColor"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                      />
                    </svg>
                  </i>
                  <i
                    className="anticon anticon-calendar ant-calendar-picker-icon"
                  >
                    <svg
                      aria-hidden="true"
                      className=""
                      data-icon="calendar"
                      fill="currentColor"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                      />
                    </svg>
                  </i>
                </div>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
  >
    <div
      className="ant-form-item-label"
    >
      <label
        className="ant-form-item-required"
        htmlFor="category"
        onClick={[Function]}
        title="Category"
      >
        Category
      </label>
    </div>
    <div
      className="ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "category",
                "value": "Auto",
              }
            }
            data-__meta={
              Object {
                "name": "category",
                "originalProps": Object {
                  "disabled": false,
                  "prefixCls": "ant-input",
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Enter a category",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Enter a category",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            disabled={false}
            id="category"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value="Auto"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={
      Object {
        "marginBottom": 5,
      }
    }
  >
    <div
      className="ant-form-item-label"
    >
      <label
        className=""
        htmlFor="notes"
        onClick={[Function]}
        title="Notes"
      >
        Notes
      </label>
    </div>
    <div
      className="ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <textarea
            className="ant-input"
            data-__field={
              Object {
                "name": "notes",
                "value": "Add notes here",
              }
            }
            data-__meta={
              Object {
                "name": "notes",
                "originalProps": Object {
                  "placeholder": "Add notes for this transaction",
                  "prefixCls": "ant-input",
                  "rows": 4,
                },
                "ref": null,
                "trigger": "onChange",
                "validate": Array [],
                "valuePropName": "value",
              }
            }
            id="notes"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Add notes for this transaction"
            rows={4}
            style={Object {}}
            value="Add notes here"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "background": "#fff",
        "borderRadius": "0 0 4px 4px",
        "borderTop": "1px solid #e8e8e8",
        "bottom": 0,
        "left": 0,
        "padding": "10px 16px",
        "position": "absolute",
        "textAlign": "right",
        "width": "100%",
      }
    }
  >
    <button
      className="ant-btn"
      onClick={[Function]}
      style={
        Object {
          "marginRight": 8,
        }
      }
      type="button"
    >
      <span>
        Cancel
      </span>
    </button>
    <button
      className="ant-btn ant-btn-primary"
      onClick={[Function]}
      type="submit"
    >
      <span>
        Submit
      </span>
    </button>
  </div>
</form>
`;
